chandra_aca/centroid_resid.py

```diff
@@ -0,0 +1,245 @@
+            yag_times = np.array(acen[ok]['time'])
+            zag_times = np.array(acen[ok]['time'])
+        elif source == 'obc':
+            telem = fetch.Msidset(['AOACYAN{}'.format(slot), 'AOACZAN{}'.format(slot)], start, stop)
+            # Filter centroids for reasonble-ness
+            yok = telem['AOACYAN{}'.format(slot)].vals != -3276.8
+            zok = telem['AOACZAN{}'.format(slot)].vals != -3276.8
+            yags = telem['AOACYAN{}'.format(slot)].vals[yok]
+            yag_times = telem['AOACYAN{}'.format(slot)].times[yok]
+            zags = telem['AOACZAN{}'.format(slot)].vals[zok]
+            zag_times = telem['AOACZAN{}'.format(slot)].times[zok]
+            raise ValueError
+        self.yags = yags
+        self.yag_times = yag_times
+        self.zags = zags
+        self.zag_times = zag_times
+
+    def set_atts(self, start, stop, source):
+        """Get attitude solution quaternions from ``source`` between
+        ``start`` and ``stop``
+
+        One could also just set atts and att_times attributes directly.
+        """
+        self.att_source = source
+        # Get attitudes and times
+        if source == 'obc':
+            # Need to figure out missing data
+            telem = fetch.Msidset(['aoattqt*'], start, stop)
+            atts = np.vstack([telem['aoattqt{}'.format(idx)].vals
+                              for idx in [1, 2, 3, 4]]).transpose()
+            att_times = telem['aoattqt1'].times
+        if source == 'ground':
+            asol_files = sorted(asp_l1.get_files(start=start, stop=stop, revision='last',
+                                                 content=['ASPSOL']))
+            acal_files = sorted(asp_l1.get_files(start=start, stop=stop, revision='last',
+                                                 content=['ACACAL']))
+            att_chunks = []
+            time_chunks = []
+            for f in asol_files:
+                asol = Table.read(f)
+                prefix = re.search("(.*)_asol1.*", f).group(1)
+                # get the acal for each asol
+                acal = None
+                # There's probably a header to check for time ranges too, but use filenames to start
+                for acf in acal_files:
+                    if re.search("{}_acal1.*".format(prefix), acf):
+                        acal = Table.read(acf)
+                        break
+                # Make a Nx4 list of the inv misalign quats
+                q_mis_inv = np.repeat(Quat(acal['aca_misalign'][0]).inv().q,
+                                      len(asol)).reshape((4, len(asol))).transpose()
+                # Quaternion multiply the asol quats with that inv misalign and save
+                # I could also do this with the transform matrix and then only need
+                # one accessory quat function.
+                att_chunks.append(quat_vmult(asol['q_att'], q_mis_inv))
+                time_chunks.append(np.array(asol['time']))
+            atts = np.vstack(att_chunks)
```

Comment:
I just wanted this separately (asol converted to ACA frame over the interval), so we might want another method to get it independently (user can also just construct a CentroidResiduals object and extract it, but that seems counterintuitive).